### PR TITLE
Scope fn

### PR DESCRIPTION
**Pull request by bwoebi**

### CODE REVIEW

File: Zend/Optimizer/zend_optimizer.c [hunk 1]
```diff
@@ -1446,23 +1474,59 @@ static void zend_redo_pass_two_ex(zend_op_array *op_array, const zend_ssa *ssa)
...
+	 * nested reservations on top, restoring T to its pre-sandwich value. */
+	if (is_scope_fn) {
+		if (op_array->T < pre_optimize_T) {
+			op_array->T = pre_optimize_T;
+		}
+		zend_refixup_scope_func_self(op_array, saved_scope_ed_offset, encoding_scope_T);
+	}
+
+	/* Redo pass_two() */
+	zend_redo_pass_two(op_array);
```
Comment (bwoebi) [inline Zend/Optimizer/zend_optimizer.c:1491]: Try move before live ranges? I think live ranges want pass two done first.

File: Zend/Optimizer/zend_optimizer.c [hunk 2]
```diff
@@ -1678,21 +1763,21 @@ ZEND_API void zend_optimize_script(zend_script *script, zend_long optimization_l
...
-		if (ZEND_OPTIMIZER_PASS_12 & optimization_level) {
-			for (i = 0; i < call_graph.op_arrays_count; i++) {
-				zend_adjust_fcall_stack_size_graph(call_graph.op_arrays[i]);
-			}
-		}
-
 		for (i = 0; i < call_graph.op_arrays_count; i++) {
 			op_array = call_graph.op_arrays[i];
+			if (zend_op_array_is_scope_fn(op_array)) {
+				continue; /* Never reverted, so don't redo */
```
Comment (bwoebi) [inline Zend/Optimizer/zend_optimizer.c]: I think we might still need to revert for scope fns for the not-skipped passes

File: Zend/Optimizer/zend_optimizer.c [hunk 3]
```diff
@@ -1669,6 +1751,9 @@ ZEND_API void zend_optimize_script(zend_script *script, zend_long optimization_l
 
 		if (ZEND_OPTIMIZER_PASS_11 & optimization_level) {
 			for (i = 0; i < call_graph.op_arrays_count; i++) {
+				if (zend_op_array_is_scope_fn(call_graph.op_arrays[i])) {
```
Comment (bwoebi) [inline Zend/Optimizer/zend_optimizer.c]: Why would we need to skip this for scope fns? Aren't parent and child literals separate? Only tmp&CV is merged

File: Zend/Optimizer/zend_optimizer.c [hunk 4]
```diff
@@ -1678,21 +1763,21 @@ ZEND_API void zend_optimize_script(zend_script *script, zend_long optimization_l
 
 		if (ZEND_OPTIMIZER_PASS_13 & optimization_level) {
 			for (i = 0; i < call_graph.op_arrays_count; i++) {
+				if (zend_op_array_is_scope_fn(call_graph.op_arrays[i])) {
```
Comment (bwoebi) [inline Zend/Optimizer/zend_optimizer.c]: Add comment why (handled in compact for nested scope fns)

File: Zend/tests/scope_fn/error_path_cleanup.phpt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+function test_recursion() {
+    $ref = null;
+    $fn = fn() {
+        $ref();
+    };
+    $ref = $fn;
+    try {
+        $fn();
+    } catch (Error) {}
+    $ref = null; // release ref before function exit to avoid escape error
```
Comment (bwoebi) [inline Zend/tests/scope_fn/error_path_cleanup.phpt]: what

File: Zend/Optimizer/zend_optimizer.c [hunk 6]
```diff
@@ -1707,6 +1782,31 @@ ZEND_API void zend_optimize_script(zend_script *script, zend_long optimization_l
...
+		 * parent's revert resets its child DECLARE.extended_value to 0.
+		 * Doing the scope-fn revert+optimize+redo cycle as one unit per
+		 * op_array, after the outer redo has finished re-encoding the
+		 * scope-fn bodies, keeps that invariant. */
+		for (i = 0; i < call_graph.op_arrays_count; i++) {
+			op_array = call_graph.op_arrays[i];
+			if (!zend_op_array_is_scope_fn(op_array)) {
+				continue;
+			}
+			zend_optimize_op_array(op_array, &ctx);
```
Comment (bwoebi) [inline Zend/Optimizer/zend_optimizer.c]: confusing continue - the inversion doesn't help

File: Zend/tests/scope_fn/undefined_parent_var.phpt [hunk 7]
```diff
@@ -0,0 +1,31 @@
+--TEST--
+Reading an undefined parent CV from a scope-fn body emits the warning with the correct variable name (regression test for crash in zval_undefined_cv)
```
Comment (bwoebi) [inline Zend/tests/scope_fn/undefined_parent_var.phpt]:
```suggestion
Reading an undefined parent CV from a scope-fn body emits the warning with the correct variable name
```

File: Zend/zend_builtin_functions.c [hunk 8]
```diff
@@ -289,6 +289,15 @@ ZEND_FUNCTION(func_get_arg)
 		RETURN_THROWS();
 	}
 
+	if (ex->func->common.fn_flags2 & ZEND_ACC2_SCOPE_FUNC) {
```
Comment (bwoebi) [inline Zend/zend_builtin_functions.c]:
```suggestion
	if (UNEXPECTED(ex->func->common.fn_flags2 & ZEND_ACC2_SCOPE_FUNC)) {
```
other places here too

File: Zend/zend_builtin_functions.c [hunk 9]
```diff
@@ -322,6 +331,26 @@ ZEND_FUNCTION(func_get_args)
...
+					ZVAL_DEREF(q);
+					Z_TRY_ADDREF_P(q);
+					ZEND_HASH_FILL_SET(q);
+				} else {
+					ZEND_HASH_FILL_SET_NULL();
+				}
+				ZEND_HASH_FILL_NEXT();
+			}
+		} ZEND_HASH_FILL_END();
+		return;
```
Comment (bwoebi) [inline Zend/zend_builtin_functions.c]: missing nNumOfElements? Merge with ZEND_HASH_FILL_PACKED below

File: Zend/zend_builtin_functions.c [hunk 10]
```diff
@@ -1714,6 +1743,32 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
...
+						ZVAL_COPY_VALUE(&original_arg, q);
+					} else {
+						ZVAL_NULL(&original_arg);
+					}
+					Z_TRY_ADDREF_P(&original_arg);
+					ZEND_HASH_FILL_SET(&original_arg);
+					ZEND_HASH_FILL_NEXT();
+				}
+			} ZEND_HASH_FILL_END();
+			return;
```
Comment (bwoebi) [inline Zend/zend_builtin_functions.c]: Same as func_get_args

File: Zend/zend_closures.c [hunk 11]
```diff
@@ -77,6 +87,10 @@ static bool zend_valid_closure_binding(
 		zend_closure *closure, zval *newthis, zend_class_entry *scope) /* {{{ */
 {
 	zend_function *func = &closure->func;
+	if (func->common.fn_flags2 & ZEND_ACC2_SCOPE_FUNC) {
```
Comment (bwoebi) [inline Zend/zend_closures.c]: We can actually allow scope changes as long as the scope is compatible?
Just $this rebinding forbidden unless identical this.
And rebinding must not create new closures. Mutate the existing one.

File: Zend/zend_compile.c [hunk 12]
```diff
@@ -8728,6 +8750,168 @@ static zend_string *zend_begin_func_decl(znode *result, zend_op_array *op_array,
...
+/* Recursive walk: apply offsets to scope_op's body, then descend into nested
+ * scope-fn children. Each nested child's frame sits at this scope_op's
+ * T_base + the child's own DECLARE_SCOPE_FUNC.extended_value. */
+static void zend_walk_scope_func_offsets(
+	zend_op_array *scope_op, uint32_t parent_last_var,
+	uint32_t T_base, uint32_t scope_T, bool encode)
+{
+	uint32_t scope_ed_offset = (uint32_t)((ZEND_CALL_FRAME_SLOT + parent_last_var + T_base + scope_T) * sizeof(zval));
+	uint32_t scope_frame_size = (uint32_t)((ZEND_CALL_FRAME_SLOT + scope_op->last_var + scope_T) * sizeof(zval));
+	zend_apply_scope_func_offsets(scope_op, scope_ed_offset, scope_frame_size, encode);
```
Comment (bwoebi) [inline Zend/zend_compile.c]: scope_frame_size can be computed from  `scope_T`, i.e. factor into function call `zend_apply_scope_func_offsets(scope_op, scope_ed_offset, scope_T, encode);`

File: Zend/zend_compile.c [hunk 13]
```diff
@@ -8844,7 +9040,11 @@ static zend_op_array *zend_compile_func_decl_ex(
 		is_method && zend_string_equals_literal(lcname, ZEND_TOSTRING_FUNC_NAME) ? IS_STRING : 0);
 	if (CG(active_op_array)->fn_flags & ZEND_ACC_GENERATOR) {
 		zend_mark_function_as_generator();
-		zend_emit_op(NULL, ZEND_GENERATOR_CREATE, NULL, NULL);
+		/* For scope functions, GENERATOR_CREATE must run AFTER ENTER_SCOPE_FUNC
+		 * (so EX is the scope_ed). It is emitted further below. */
+		if (!is_scope_fn) {
+			zend_emit_op(NULL, ZEND_GENERATOR_CREATE, NULL, NULL);
+		}
 	}
```
Comment (bwoebi) [inline Zend/zend_compile.c:9052]: We can reorder this below the big is_scope_fn branch

File: Zend/zend_execute.c [hunk 14]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+ZEND_API void zend_vm_stack_free_tracked_temporaries_ex(zend_execute_data *call)
+{
+	const zend_op_array *op_array = &call->func->op_array;
+	zval *base = ZEND_CALL_VAR_NUM(call, op_array->last_var + op_array->T - 1);
+	uint32_t count = Z_EXTRA_P(base) >> 8;
+
+	for (uint32_t i = 0; i < count; i++) {
+		zval *entry = base - i - 1;
+		ZEND_ASSERT((Z_EXTRA_P(entry) & 0xFF) == ZEND_TRACKED_TMP_SCOPE_FUNC);
+		zval *tmp = zend_closure_get_this_ptr_ptr(Z_OBJ_P(entry));
```
Comment (bwoebi) [inline Zend/zend_execute.c]: Oddly undescriptive name

File: Zend/zend_execute.c [hunk 15]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+		bool escaped = GC_REFCOUNT(Z_OBJ_P(entry)) > 1;
+
+		if (escaped) {
```
Comment (bwoebi) [inline Zend/zend_execute.c]:
```suggestion
		if (UNEXPECTED(escaped)) {
```

File: Zend/zend_execute.c [hunk 16]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+			 && (*attached_object_ptr)->ce == zend_ce_generator) {
+				zend_generator *gen = (zend_generator *) *attached_object_ptr;
+				zend_generator_close(gen, false);
+			}
+			ZEND_ASSERT(*zend_closure_get_attached_object_ptr(Z_OBJ_P(entry)) == NULL);
+		}
+
+		Z_PTR_P(tmp) = NULL; /* invalidate parent_execute_data */
+		OBJ_RELEASE(Z_OBJ_P(entry));
+		if (escaped && EG(exception) == NULL) {
```
Comment (bwoebi) [inline Zend/zend_execute.c]: this can be included into the branch above

File: Zend/zend_execute.c [hunk 17]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+					zend_fiber *fiber = (zend_fiber *)gen->fiber_running_me;
+					if (fiber->context.status == ZEND_FIBER_STATUS_SUSPENDED
+					 && fiber->caller == NULL) {
+						unwind_fiber = fiber;
+					}
+				}
+			}
+
+			if (unwind_fiber != NULL) {
+				zend_fiber *fiber = unwind_fiber;
```
Comment (bwoebi) [inline Zend/zend_execute.c]: unnecessary assignment

File: Zend/zend_execute.c [hunk 18]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+				 * boundary. Held in deferred_exception so
+				 * zend_fiber_block_suspend_for_forced_unwind can
+				 * re-inject it if a finally block calls Fiber::suspend. */
+				zend_object *exc = zend_create_scope_fn_unwind();
+				GC_ADDREF(exc);
+				fiber->deferred_exception = exc;
+				fiber->forced_unwind_target = target;
+
+				zval thrown;
+				ZVAL_OBJ(&thrown, exc); /* steals the remaining ref */
```
Comment (bwoebi) [inline Zend/zend_execute.c]: Strip the addref, use ZVAL_OBJ_COPY here

File: Zend/zend_execute.c [hunk 19]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+				 * up — we must dtor it ourselves below. */
+				uint8_t *undef_before = NULL;
+				uint32_t cv_count = op_array->last_var;
+				if (cv_count > 0) {
+					undef_before = ecalloc(cv_count, sizeof(uint8_t));
+					for (uint32_t v = 0; v < cv_count; v++) {
+						zval *cv = ZEND_CALL_VAR_NUM(call, v);
+						undef_before[v] = (Z_TYPE_P(cv) == IS_UNDEF);
+					}
+				}
```
Comment (bwoebi) [inline Zend/zend_execute.c]: What. How is this supposed to be possible that CVs are assigned at all during this cleanup?

File: Zend/zend_vm_def.h [hunk 20]
```diff
@@ -2992,6 +2992,88 @@ ZEND_VM_HOT_HELPER(zend_leave_helper, ANY, ANY)
...
+			if (UNEXPECTED(unwind_fiber != NULL)) {
+				/* TOP: scope_fn body is the fiber body. Suspend before the
+				 * natural fiber-finish so the user can resume and have the
+				 * Error injected. The injection lands here on resume; the
+				 * subsequent ZEND_VM_RETURN unwinds the fiber body with the
+				 * Error in flight, surfacing it through Fiber::resume's
+				 * transfer. */
+				zend_fiber_synthetic_suspend(unwind_fiber, NULL);
+			}
+			ZEND_VM_RETURN();
```
Comment (bwoebi) [inline Zend/zend_vm_def.h]: This code path skips zend_scope_ed_pop_original_call_frame?

Seems likely to leak memory. You'll have to always call this before. We should have a test for this case.

File: Zend/zend_execute.c [hunk 21]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+							}
+						}
+					}
+					efree(undef_before);
+				}
+
+				if (transfer.flags & ZEND_FIBER_TRANSFER_FLAG_ERROR) {
+					/* Don't re-raise the sentinel above the scope-fn —
+					 * it's an internal marker, not user-visible. */
+					if (zend_is_scope_fn_unwind(Z_OBJ(transfer.value))) {
```
Comment (bwoebi) [inline Zend/zend_execute.c]: the unwind-exception should always have been cleared in leave_helper. This should never be true and rather be an assert.

File: Zend/zend_execute.c [hunk 22]
```diff
@@ -4973,6 +4991,166 @@ ZEND_API void zend_cleanup_unfinished_execution(zend_execute_data *execute_data,
...
+					zval_ptr_dtor(&transfer.value);
+					EG(exception) = prev_exception;
+				}
+			}
+
+			/* Whatever was attached, finalize it. The fiber-attached
+			 * unwind detached the closure via leave_helper. The
+			 * generator-attached case (with or without a fiber inside)
+			 * still needs its generator closed. */
+			if (*attached_object_ptr != NULL
```
Comment (bwoebi) [inline Zend/zend_execute.c]: shouldn't that be else if? when the generator was in the fiber, the unwind in the fiber should already have closed it.

File: Zend/zend_execute.h [hunk 23]
```diff
@@ -331,6 +332,48 @@ static zend_always_inline zend_vm_stack zend_vm_stack_new_page(size_t size, zend
...
+
+/* True iff `ex` is a scope-fn frame. The tag bit is only set by
+ * ZEND_ENTER_SCOPE_FUNC, but extra_named_params is not zero-initialized
+ * for every frame in the engine — gate on the function flag first to
+ * avoid reading a tag bit out of garbage. */
+static zend_always_inline bool zend_is_scope_ed(const zend_execute_data *ex)
+{
+	return ex->func != NULL
+		&& (ex->func->common.fn_flags2 & ZEND_ACC2_SCOPE_FUNC)
+		&& ((uintptr_t)ex->extra_named_params & ZEND_SCOPE_ED_ENP_TAG_SCOPE_ED);
```
Comment (bwoebi) [inline Zend/zend_execute.h]: Is this check for ZEND_SCOPE_ED_ENP_TAG_SCOPE_ED even necessary? ZEND_ACC2_SCOPE_FUNC should be enough, right?

File: Zend/zend_execute.h [hunk 24]
```diff
@@ -393,6 +440,25 @@ static zend_always_inline void zend_vm_stack_free_extra_args(zend_execute_data *
...
+ *
+ * Guarded by ZEND_CALL_TRACKED_TEMPORARIES flag (shared with ZEND_CALL_FREE_EXTRA_ARGS)
+ * and ZEND_ACC2_HAS_TRACKED_TEMPORARIES in fn_flags2. */
+#define ZEND_TRACKED_TMP_SCOPE_FUNC 2
+
+ZEND_API void zend_vm_stack_free_tracked_temporaries_ex(zend_execute_data *execute_data);
+
+static zend_always_inline void zend_vm_stack_free_tracked_temporaries(uint32_t call_info, zend_execute_data *execute_data)
+{
+	if (UNEXPECTED((call_info & ZEND_CALL_TRACKED_TEMPORARIES) != 0) && ZEND_USER_CODE(EX(func)->type) && (EX(func)->op_array.fn_flags2 & ZEND_ACC2_HAS_TRACKED_TEMPORARIES)) {
```
Comment (bwoebi) [inline Zend/zend_execute.h]: ZEND_USER_CODE(EX(func)->type)

drop this, no need to force this only for user functions here - fn_flags check is enough (via EX(func)->common.fn_flags)

File: Zend/zend_execute.h [hunk 25]
```diff
@@ -447,6 +513,75 @@ static zend_always_inline void zend_vm_stack_extend_call_frame(
...
+			zend_free_extra_named_params(original_call_frame->extra_named_params);
+		}
+		EG(vm_stack_top) = (zval *)original_call_frame;
+	}
+}
+
+/* Find the literal-pool index where a scope-fn op_array's parameter→parent-CV
+ * mapping starts. Stored on ENTER_SCOPE_FUNC's op2.num at compile time. The
+ * scan is bounded by num_args + a few prologue opcodes. */
+static zend_always_inline uint32_t zend_scope_fn_first_literal(const zend_op_array *op_array)
```
Comment (bwoebi) [inline Zend/zend_execute.h]: Rather than a lookup, can we always put arg->CV mapping literals starting at index 0? Should simplify some tthings.

Should be easy to do in compiler - allocate as many literals as we have args early in zend_compile_func_decl_ex?

File: Zend/zend_execute_API.c [hunk 26]
```diff
@@ -1806,6 +1806,22 @@ ZEND_API zend_array *zend_rebuild_symbol_table(void) /* {{{ */
 	if (!ex) {
 		return NULL;
 	}
+
+	/* Scope functions share CVs with their parent — rebuild on the parent
+	 * frame and adopt the result. */
+	if (ex->func->common.fn_flags2 & ZEND_ACC2_SCOPE_FUNC) {
```
Comment (bwoebi) [inline Zend/zend_execute_API.c]: Why this branch before `if (ZEND_CALL_INFO(ex) & ZEND_CALL_HAS_SYMBOL_TABLE) {`?

File: Zend/zend_generators.c [hunk 27]
```diff
@@ -864,6 +928,17 @@ ZEND_API void zend_generator_resume(zend_generator *orig_generator) /* {{{ */
...
+	/* Scope-fn unwind sentinel: thrown into the body when the closure's
+	 * declaring scope is exiting. Finally blocks have already run via the
+	 * normal exception machinery; absorb the sentinel here so it never
+	 * surfaces to user code. The generator was already closed at the
+	 * source of the unwind. */
+	if (UNEXPECTED(EG(exception) != NULL)
+	 && zend_is_scope_fn_unwind(EG(exception))) {
+		OBJ_RELEASE(EG(exception));
+		EG(exception) = NULL;
+	}
```
Comment (bwoebi) [inline Zend/zend_generators.c]: Merge into branch below

File: Zend/zend_generators.c [hunk 28]
```diff
@@ -852,6 +912,10 @@ ZEND_API void zend_generator_resume(zend_generator *orig_generator) /* {{{ */
 		}
 	}
 	generator->flags &= ~(ZEND_GENERATOR_CURRENTLY_RUNNING | ZEND_GENERATOR_IN_FIBER);
+	generator->fiber_running_me = NULL;
+	if (orig_generator != generator) {
+		orig_generator->fiber_running_me = NULL;
+	}
```
Comment (bwoebi) [inline Zend/zend_generators.c]: This belongs right at the last line of this function next to `orig_generator->flags &= ~(ZEND_GENERATOR_DO_INIT | ZEND_GENERATOR_IN_FIBER);`.

File: Zend/zend_vm_def.h [hunk 29]
```diff
@@ -10639,3 +10804,179 @@ ZEND_VM_HELPER(zend_interrupt_helper, ANY, ANY)
...
+	 * which CV each param maps to). The scope_ed frame has no own arg
+	 * slots — the body reads its parameters via parent CVs. */
+	if (num_params > 0) {
+		uint32_t first_literal = opline->op2.num;
+		zval *literals = EX(func)->op_array.literals;
+		for (uint32_t i = 0; i < num_params; i++) {
+			zval *src = ZEND_CALL_ARG(call_frame, i + 1);
+			uint32_t parent_cv_offset = (uint32_t)Z_LVAL(literals[first_literal + i]);
+			zval *dst = ZEND_CALL_VAR(parent_ed, parent_cv_offset);
+			zval_ptr_dtor(dst);
```
Comment (bwoebi) [inline Zend/zend_vm_def.h]: destroying dst this way is unsafe due to destructors.
You need to first replace it with the new value, then you can free it.

File: Zend/zend_vm_def.h [hunk 30]
```diff
@@ -9725,6 +9866,30 @@ ZEND_VM_HANDLER(172, ZEND_FUNC_GET_ARGS, UNUSED|CONST, UNUSED)
...
+				zval *q = zend_scope_fn_get_arg_zval(execute_data, i, first_literal);
+				if (q && EXPECTED(Z_TYPE_INFO_P(q) != IS_UNDEF)) {
+					ZVAL_DEREF(q);
+					Z_TRY_ADDREF_P(q);
+					ZEND_HASH_FILL_SET(q);
+				} else {
+					ZEND_HASH_FILL_SET_NULL();
+				}
+				ZEND_HASH_FILL_NEXT();
+			}
```
Comment (bwoebi) [inline Zend/zend_vm_def.h:9871]: I don't see any extra_args handling here??

File: Zend/zend_vm_def.h
Comment (bwoebi) [inline Zend/zend_vm_def.h]: specialize via `SPEC(SCOPE_FN)`

File: Zend/zend_vm_def.h [hunk 31]
```diff
@@ -4723,19 +4832,51 @@ ZEND_VM_HANDLER(139, ZEND_GENERATOR_CREATE, ANY, ANY)
...
+			/* scope_ed lives in parent's TMP, not heap-allocated. */
+			ZEND_ADD_CALL_FLAG_EX(call_info, ZEND_CALL_ALLOCATED);
+		}
 		Z_TYPE_INFO(gen_execute_data->This) = call_info;
-		gen_execute_data->prev_execute_data = NULL;
+		if (!is_scope_fn) {
+			/* scope-fn keeps prev_execute_data — body CV access traverses it. */
+			gen_execute_data->prev_execute_data = NULL;
+		}
+
```
Comment (bwoebi) [inline Zend/zend_vm_def.h:4836]: merge branches

File: Zend/zend_vm_def.h [hunk 32]
```diff
@@ -10639,3 +10804,179 @@ ZEND_VM_HELPER(zend_interrupt_helper, ANY, ANY)
...
+	ZVAL_COPY_VALUE(&scope_ed->This, &EX(This));
+	ZEND_CALL_NUM_ARGS(scope_ed) = ZEND_CALL_NUM_ARGS(call_frame);
+
+	Z_EXTRA_P(this_ptr) = 1; /* recursion guard */
+
+	/* Transfer (move) the ZEND_CALL_CLOSURE ref from INIT_DYNAMIC_CALL to
+	 * scope_ed — no extra ADDREF; leave_helper releases it. */
+
+	execute_data = scope_ed;
+	EG(current_execute_data) = scope_ed;
```
Comment (bwoebi) [inline Zend/zend_vm_def.h]: transferring here will ignore updates to scope_ed->opline when there was an exception during arg receive.

I think we need to move arg->CV remapping beyond this line to avoid this issue. (Add a test where an object with a throwing destructor gets overwritten by received args.)

File: Zend/zend_vm_def.h [hunk 33]
```diff
@@ -2992,6 +2992,88 @@ ZEND_VM_HOT_HELPER(zend_leave_helper, ANY, ANY)
 	SAVE_OPLINE();
 #endif
 
+	/* Scope function: don't free CVs, don't pop VM stack for scope_ed itself —
+	 * the parent owns them. */
+	if (UNEXPECTED(zend_is_scope_ed(execute_data))) {
```
Comment (bwoebi) [inline Zend/zend_vm_def.h]: Big no-no to introducing such an expensive EX(func) check in the main hot path of zend_leave_helper.
I just discovered that there exists ZEND_CALL_OBSERVED. And we DO need to set it, otherwise the optimization in ZEND_RETURN will overwrite our CV! (add a test case for this - returning a refcounted CV in a scope fn must not null it)
Let's reuse this call flag and mark any scope fns ZEND_CALL_SCOPE_FN (aliased to ZEND_CALL_OBSERVED).
Now we can avoid this on the main hot path, and instead add this check in `if (EXPECTED((call_info & ZEND_CALL_CODE) == 0)) {` and `if (EXPECTED((call_info & (ZEND_CALL_CODE|ZEND_CALL_TOP)) == 0)`.

File: Zend/zend_execute.h [hunk 34]
```diff
@@ -447,6 +513,75 @@ static zend_always_inline void zend_vm_stack_extend_call_frame(
...
+{
+	if (original_call_frame) {
+		uint32_t orig_info = ZEND_CALL_INFO(original_call_frame);
+		if (UNEXPECTED(orig_info & ZEND_CALL_FREE_EXTRA_ARGS)) {
+			zend_vm_stack_free_extra_args_ex(orig_info, original_call_frame);
+		}
+		if (UNEXPECTED(orig_info & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+			zend_free_extra_named_params(original_call_frame->extra_named_params);
+		}
+		EG(vm_stack_top) = (zval *)original_call_frame;
```
Comment (bwoebi) [inline Zend/zend_execute.h]: Don't we need to handle ZEND_CALL_ALLOCATED here?

File: Zend/zend_execute.h [hunk 35]
```diff
@@ -447,6 +513,75 @@ static zend_always_inline void zend_vm_stack_extend_call_frame(
 
 ZEND_API void ZEND_FASTCALL zend_free_extra_named_params(zend_array *extra_named_params);
 
+/* Pop the call frame that originally invoked a scope fn from the vm_stack.
+ * `original_call_frame` is recovered from a scope_ed's extra_named_params via
+ * the SCOPE_ED_ENP_TAG_MASK clear. Frees extra args / extra named params on
+ * that frame, then resets vm_stack_top. */
+static zend_always_inline void zend_scope_ed_pop_original_call_frame(zend_execute_data *original_call_frame)
+{
+	if (original_call_frame) {
```
Comment (bwoebi) [inline Zend/zend_execute.h]: Is this ever NULL or why this check?

File: Zend/zend_vm_def.h [hunk 36]
```diff
@@ -4723,19 +4832,51 @@ ZEND_VM_HANDLER(139, ZEND_GENERATOR_CREATE, ANY, ANY)
...
+			*attached_object_ptr = &generator->std;
+
+			uintptr_t enp_tag = (uintptr_t)execute_data->extra_named_params;
+			zend_execute_data *original_call_frame =
+				(zend_execute_data *)(enp_tag & ~(uintptr_t)ZEND_SCOPE_ED_ENP_TAG_MASK);
+			execute_data->extra_named_params = (zend_array *)(enp_tag | ZEND_SCOPE_ED_ENP_TAG_OBJECT_ATTACHED);
+
+			EG(current_execute_data) = EX(prev_execute_data);
+			execute_data = EX(prev_execute_data);
+			if (UNEXPECTED(orig_call_info & ZEND_CALL_TOP)) {
```
Comment (bwoebi) [inline Zend/zend_vm_def.h:4851]: This codepath skips zend_scope_ed_pop_original_call_frame (to free extra args etc.)?

File: Zend/zend_vm_def.h (hunk 33 above)
Comment (bwoebi) [reply Zend/zend_vm_def.h]: And the main part of this code can become an inline function in zend_execute.c.

File: Zend/zend_vm_def.h [hunk 37]
```diff
@@ -3032,6 +3114,7 @@ ZEND_VM_HOT_HELPER(zend_leave_helper, ANY, ANY)
 		/* Free extra args before releasing the closure,
 		 * as that may free the op_array. */
 		zend_vm_stack_free_extra_args_ex(call_info, execute_data);
+		zend_vm_stack_free_tracked_temporaries(call_info, execute_data);
```
Comment (bwoebi) [inline Zend/zend_vm_def.h:3042]: We should merge this. `zend_vm_stack_free_extra_args_and_tracked_temporaries`.

File: Zend/zend_fibers.c [hunk 38]
```diff
@@ -955,6 +1019,25 @@ ZEND_METHOD(Fiber, suspend)
...
+/* If a parent's exit force-unwound this fiber, the visible escape Error was
+ * stashed on the fiber inside the scope_fn boundary handler. Inject it as
+ * the in-flight exception of the resumed fiber so the throw materializes at
+ * the suspension resumption point inside the fiber body — any try/catch
+ * surrounding the suspending call there will see it.
+ *
+ * Returns a zval* containing the exception value to pass to resume_internal
+ * if a deferred throw is pending; otherwise NULL. The caller forwards this
+ * zval as the resume value with `exception=true`. */
+static zval *zend_fiber_consume_pending_resume_throw(zend_fiber *fiber, zval *out)
```
Comment (bwoebi) [inline Zend/zend_fibers.c]: weird function, would be easier to reason about if this code were inline and not its own function.

File: Zend/zend_fibers.c [hunk 39]
```diff
@@ -747,8 +748,63 @@ ZEND_API void zend_fiber_resume_exception(zend_fiber *fiber, zval *exception, zv
...
+}
+
 ZEND_API void zend_fiber_suspend(zend_fiber *fiber, zval *value, zval *return_value)
 {
+	if (zend_fiber_block_suspend_for_forced_unwind(fiber)) {
+		if (return_value) {
+			ZVAL_NULL(return_value);
+		}
+		return;
+	}
```
Comment (bwoebi) [inline Zend/zend_fibers.c]: Actually, we should temporarily set `ZEND_FIBER_FLAG_DESTROYED` while we're unwinding up to a scope fn (which will make Fiber::suspend throw, but that's what's wanted). Instead of returning NULL.
